sympy/solvers/polysys.py

```diff
@@ -78,6 +78,147 @@ def solve_poly_system(seq, *gens, strict=False, **args):
+        sets_without_A = factors_system(all_eqs_A_removed)
+
+        return sets_with_A + sets_without_A
+
+    def compute_irreducible_groebner_bases(
+        eqs: List[Poly],
+        syms: Tuple[Symbol, ...],
+        order: str = 'lex',
+        method: str = 'buchberger',
+        modulus: int = None,
+        **kwargs
+    ) -> List[List[Poly]]:
+
+        def process_system(system: List[Poly]) -> List[List[Poly]]:
+            gb = groebner(system, syms, order=order, method=method, modulus=modulus, **kwargs)
+            factored_system = factor_poly_system(gb, modulus=modulus)
+
+            if len(factored_system) > 1:
+                subsystems = []
+                for subsystem in factored_system:
+                    subsystems.extend(process_system(subsystem))
+                return subsystems
+            else:
+                return [[simplify(g) for g in gb]]
+
+        def remove_redundant_systems(systems: List[List[Poly]]) -> List[List[Poly]]:
+            unique_systems = []
+            for system in systems:
+                if not any(set(system).issubset(set(existing_system)) for existing_system in unique_systems):
+                    unique_systems.append(system)
+            return unique_systems
+
+        irreducible_systems = process_system(eqs)
+        return remove_redundant_systems(irreducible_systems)
+
+    def solve_polynomial_system(
+        eqs: List[Poly],
+        syms: Tuple[Symbol, ...],
+        order: str = 'lex',
+        method: str = 'buchberger',
+        modulus: int = None,
+        **kwargs
+    ) -> List[Dict[Symbol, Any]]:
+
+        irreducible_systems = compute_irreducible_groebner_bases(eqs, syms, order, method, modulus, **kwargs)
+        solutions = []
+
+        for system in irreducible_systems:
+            if all(poly.is_univariate for poly in system):
+                solutions.extend(solve(system, syms, dict=True))
+            else:
+                sub_gb = groebner(system, syms, order=order, method=method, modulus=modulus, **kwargs)
+                solved_symbols = []
+                solutions_dict = {}
+                unspecified_symbols = list(syms)
+
+                for poly in reversed(sub_gb):
+                    for symbol in reversed(syms):
+                        if symbol in poly.free_symbols:
+                            if all(s not in poly.free_symbols for s in syms[:syms.index(symbol)]):
+                                solved_symbols.append(symbol)
+                                solutions_dict[symbol] = poly
+                                unspecified_symbols.remove(symbol)
+                                break
+
+                if not solved_symbols:
+                    continue
+
+                solved_symbols.reverse()
+
+                determined_system = [solutions_dict[sym] for sym in solved_symbols]
+                determined_solution = solve(determined_system, *solved_symbols, dict=True)
```

Comment:
We shouldn't call `solve` here. This function would eventually be called as part of solve.